From 367c188b2f4317e9ea069ba5e7b4cc0178466cca Mon Sep 17 00:00:00 2001
From: "Tekriwal, Tanuj" <tanuj.tekriwal@intel.corp-partner.google.com>
Date: Mon, 8 Jun 2026 09:34:36 +0000
Subject: [PATCH] InputDevice: Use location for idc file

Devices that have multiple usb touchpanels attached that are all using
the same product, vendor, and version ids do not have a good way to use
different idc files to allow the touchpanels to be configured
differently as matching is done either by product, vendor, and version
ids or device name which will be the same for all of the touchpanels.

This updates the logic to search for idc files to add searching by
InputDeviceIdentifier.location if no existing matches are found. The
location is an identifier such as usb-xhci-hcd.1.auto-1/input1 which
contains a bus id to allow configuration by which usb port the
touchpanel is attached to. This will use the canonical location which
replaces anything other than A-Z, a-z, 0-9, -, and _ with the _
character similar to what is done for the canonical name to allow it
to be used in a path.

Tests: Verified 2 touch display by mapping location based idc files to
display.

Type: New Feature
Change-Id: I30db47889f4925e8eb6522640bd8316af20ca572

Tracked-On: OAM-104623
Change-Id: Ib1695ac1bbc32a73bd37fe7ae9323274901d4495
Signed-off-by: Vilas R K <vilas.r.k@intel.com>
---
 include/input/InputDevice.h |  9 +++++++++
 libs/input/InputDevice.cpp  | 26 ++++++++++++++++++++++++--
 2 files changed, 33 insertions(+), 2 deletions(-)

diff --git a/include/input/InputDevice.h b/include/input/InputDevice.h
index b662b1311d..de7dbba491 100644
--- a/include/input/InputDevice.h
+++ b/include/input/InputDevice.h
@@ -71,6 +71,15 @@ struct InputDeviceIdentifier {
      */
     std::string getCanonicalName() const;
 
+    /**
+     * Return InputDeviceIdentifier.location that has been adjusted as follows:
+     *     - all characters besides alphanumerics, dash,
+     *       and underscore have been replaced with underscores.
+     * This helps in situations where a file that matches the device location is needed,
+     * while conforming to the filename limitations.
+     */
+    std::string getCanonicalLocation() const;
+
     bool operator==(const InputDeviceIdentifier&) const = default;
     bool operator!=(const InputDeviceIdentifier&) const = default;
 };
diff --git a/libs/input/InputDevice.cpp b/libs/input/InputDevice.cpp
index a768c95ede..d05bffd8fb 100644
--- a/libs/input/InputDevice.cpp
+++ b/libs/input/InputDevice.cpp
@@ -92,8 +92,20 @@ std::string getInputDeviceConfigurationFilePathByDeviceIdentifier(
     }
 
     // Try device name.
-    return getInputDeviceConfigurationFilePathByName(deviceIdentifier.getCanonicalName() + suffix,
-                                                     type);
+    std::string productPath =
+    getInputDeviceConfigurationFilePathByName(deviceIdentifier.getCanonicalName(),
+                                              type);
+    if (!productPath.empty()) {
+        return productPath;
+    }
+
+    // Try device location.
+    if (!deviceIdentifier.getCanonicalLocation().empty()) {
+        productPath = getInputDeviceConfigurationFilePathByName(
+            deviceIdentifier.getCanonicalLocation(), type);
+    }
+    return productPath;
+
 }
 
 std::string getInputDeviceConfigurationFilePathByName(
@@ -179,6 +191,16 @@ std::string InputDeviceIdentifier::getCanonicalName() const {
     return replacedName;
 }
 
+std::string InputDeviceIdentifier::getCanonicalLocation() const {
+    std::string replacedLocation = location;
+    for (char& ch : replacedLocation) {
+        if (!isValidNameChar(ch)) {
+            ch = '_';
+        }
+    }
+    return replacedLocation;
+}
+
 // --- InputDeviceInfo ---
 InputDeviceInfo::InputDeviceInfo() {
     initialize(/*id=*/-1, /*generation=*/0, /*controllerNumber=*/-1, InputDeviceIdentifier(),
-- 
2.34.1

